utils/near-rate-limiter/src/framed_read.rs

```diff
@@ -0,0 +1,380 @@
+use std::borrow::BorrowMut;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::{io, task};
+
+use bytes::BytesMut;
+use futures_core::{ready, Stream};
+use log::trace;
+
+use pin_project_lite::pin_project;
+use tokio::io::AsyncRead;
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+use tokio_util::io::poll_read_buf;
+
+// Initial capacity of read buffer.
+const INITIAL_CAPACITY: usize = 8 * 1024;
+
+// Max number of messages we received from peer, and they are in progress, before we start throttling.
+const MAX_MESSAGES_COUNT: usize = 20;
+// Max total size of all messages that are in progress, before we start throttling.
+const MAX_MESSAGES_TOTAL_SIZE: usize = 500_000_000;
+
+pin_project! {
+    pub struct ThrottledFrameRead<T, D> {
+        #[pin]
+        inner: FramedImpl<T, D, ReadFrame>,
+    }
+}
+
+pin_project! {
+    #[derive(Debug)]
+    pub(crate) struct FramedImpl<T, U, State> {
+        #[pin]
+        pub(crate) inner: T,
+        pub(crate) state: State,
+        pub(crate) codec: U,
+        pub(crate) rate_limiter: RateLimiterHelper,
+        pub(crate) receiver: UnboundedReceiver<()>,
+    }
+}
+
+// Wrapper around Actix messages, used to track size of all messages sent to PeerManager.
+// TODO(#5155) Finish implementation of this.
+
+#[allow(unused)]
+/// TODO - Once we start using this `ActixMessageWrapper` we will need to make following changes
+/// to get this struct to work
+/// - Add needed decorators. Probably `Debug`, `Message` from Actix, etc.
+/// - Add two rate limiters (local per peer, global one)
+/// - Any other metadata we need debugging, etc.
+pub struct ActixMessageWrapper<T> {
+    msg: T,
+    rate_limiter: RateLimiterHelper,
+    msg_len: usize,
+}
+
+impl<T> ActixMessageWrapper<T> {
+    #[allow(unused)]
+    pub fn new(msg: T, rate_limiter: RateLimiterHelper) -> Self {
+        // TODO(#5155) Add decorator like SizeOf
+        let msg_len = 0; // TODO msg.sizeof()
+        rate_limiter.add_msg(msg_len);
+        Self { msg, rate_limiter, msg_len }
+    }
+
+    #[allow(unused)]
+    pub fn take(mut self) -> (T, RateLimiterHelper) {
+        self.rate_limiter.remove_msg(self.msg_len);
+
+        return (self.msg, self.rate_limiter);
+    }
+}
+
+/// Each RateLimiterHelper is associated with exactly one `Peer`. It keeps track total size of all
+/// messages / their count that were received from given peer. We are going to keep tracking them
+/// as long as their exist in transit in inside `Actix` queues. For example from `Peer` to
+/// `PeerManager` and back to `Peer`.
+///
+/// TODO (#5155) Add throttling by bandwidth.
+#[derive(Clone, Debug)]
+pub struct RateLimiterHelper {
```

Comment:
Not a huge fan of the naming:

* RateLimiter is inconsistent with throttling terminology we use elsewhere 
* `Helper` suffix adds zero bits of information

Don't know a good, catchy name here, but something unimaginative like `ThrottleController` might work.

Reply:
Ok, `ThrottleController` is fine with me.